dev/io.openliberty.microprofile.telemetry.internal.common/src/io/openliberty/microprofile/telemetry/internal/common/info/OpenTelemtryLifecycleManagerImpl.java

```diff
@@ -61,38 +60,20 @@ public OpenTelemtryLifecycleManagerImpl(@Reference MetaDataSlotService slotServi
         //the runtimeMode was added in 2.0
         final boolean telemetry2OrLater = shortName.startsWith("mpTelemetry-1") ? false : true;
 
-        CheckpointHook checkpointHook = new CheckpointHook() {
-
-            @Override
-            public void restore() {
-                waitingForCheckpointRestore = false;
-                checkThenSetRuntimeFields(telemetry2OrLater);
-            }
-        };
-
-        // If checkpoint is enabled this statement will register the hook we created above. Then checkpoint returns true so the if statement is false.
-        // If checkpoint is disabled it does nothing and returns false, so the if statement will be true.
-        if (!CheckpointPhase.getPhase().addMultiThreadedHook(checkpointHook)) {
-            //Checkpoint disabled
+        //This code runs immediately if checkpoint is disabled. Otherwise it runs on checkpoint restore
+        CheckpointPhase.onRestore(() -> {
             waitingForCheckpointRestore = false;
-            checkThenSetRuntimeFields(telemetry2OrLater);
-        } else {
-            //Checkpoint enabled, before restore.
-            waitingForCheckpointRestore = true;
-        }
-    }
-
-    private void checkThenSetRuntimeFields(boolean telemetry2OrLater) {
-        HashMap<String, String> propreties = OpenTelemetryPropertiesReader.getRuntimeInstanceTelemetryProperties();
-        isRuntimeEnabled = telemetry2OrLater && !!!OpenTelemetryPropertiesReader.checkDisabled(propreties);
+            HashMap<String, String> propreties = OpenTelemetryPropertiesReader.getRuntimeInstanceTelemetryProperties();
+            isRuntimeEnabled = telemetry2OrLater && OpenTelemetryPropertiesReader.isEnabled(propreties);
 
-        if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
-            Tr.debug(this, tc, "Configured runtime mode as ", isRuntimeEnabled);
-        }
+            if (isRuntimeEnabled) {
+                runtimeInstance = LazyInitializer.<OpenTelemetryInfo> builder().setInitializer(curryInfoFactory(isRuntimeEnabled)).get();
```

Comment:
The three fields `waitingForCheckpointRestore`, `isRuntimeEnabled` and `runtimeInstance` should be marked volatile since no synchronization is used to set/read them.  Focusing on the order they are read in `getOpenTelemetryInfo` it seems this method should set the final values for these in the reverse order.  Otherwise it is possible to get an NPE for the following block if another thread tries to call `runtimeInstance.get()` before the lazy initialization of the object is assigned to `runtimeInstance` here after you already set `isRuntimeEnabled` to true:

```java
        if (isRuntimeEnabled) {
            if (TraceComponent.isAnyTracingEnabled() && tc.isDebugEnabled()) {
                Tr.debug(tc, "Returning {0} OTEL instance.", OpenTelemetryConstants.OTEL_RUNTIME_INSTANCE_NAME);
            }
            try {
                return runtimeInstance.get();
```